### PR TITLE
Add PhiOperand and PhiValue types. 

### PR DESCRIPTION
**Pull request by atrick**
Ignore the first commit. This is on top of:
https://github.com/apple/swift/pull/35676

This is necessitated by the SILArgument representation. It has the
tragic property that adding unrelated phis invalidates existing
phis. Therefore, the optimizer can't do book-keeping of phi values by
refering directly to SILValues or Operands. Instead, it must only
refer to SILBasicBlocks and argument indices.

### CODE REVIEW

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci test

File: include/swift/SIL/SILBasicBlock.h [hunk 1]
```diff
@@ -488,4 +488,126 @@ struct ilist_traits<::swift::SILBasicBlock>
...
+    auto *branch = dyn_cast<BranchInst>(operand->getUser());
+    if (!branch)
+      return;
+
+    predBlock = branch->getParent();
+    argIndex = operand->getOperandNumber();
+  }
+
+  bool operator==(PhiOperand other) const {
+    return predBlock == other.predBlock && argIndex == other.argIndex;
```
Comment (gottesmm) [inline include/swift/SIL/SILBasicBlock.h:525]: Can you add a != in a subsequent commit?

File: include/swift/SIL/SILBasicBlock.h [hunk 2]
```diff
@@ -488,4 +488,126 @@ struct ilist_traits<::swift::SILBasicBlock>
...
+    auto *blockArg = dyn_cast<SILPhiArgument>(value);
+    if (!blockArg || !blockArg->isPhiArgument())
+      return;
+
+    phiBlock = blockArg->getParent();
+    argIndex = blockArg->getIndex();
+  }
+
+  bool operator==(PhiValue other) const {
+    return phiBlock == other.phiBlock && argIndex == other.argIndex;
```
Comment (gottesmm) [inline include/swift/SIL/SILBasicBlock.h:563]: Also here.

File: include/swift/SIL/SILBasicBlock.h [hunk 3]
```diff
@@ -488,4 +488,126 @@ struct ilist_traits<::swift::SILBasicBlock>
...
+  }
+  static unsigned getHashValue(swift::PhiOperand phiOper) {
+    return llvm::hash_combine(phiOper.predBlock, phiOper.argIndex);
+  }
+  static bool isEqual(swift::PhiOperand lhs, swift::PhiOperand rhs) {
+    return lhs == rhs;
+  }
+};
+
+template <> struct DenseMapInfo<swift::PhiValue> {
```
Comment (gottesmm) [inline include/swift/SIL/SILBasicBlock.h:595]: Can you rebase this so you get my change to .clang-format +use git-clang-format so template is on the next line. = ).

File: lib/SIL/Utils/OwnershipUtils.cpp [hunk 4]
```diff
@@ -878,6 +845,160 @@ OwnedValueIntroducer swift::getSingleOwnedValueIntroducer(SILValue inputValue) {
...
+    case OperandOwnership::PointerEscape:
+      return false;
+
+    case OperandOwnership::InstantaneousUse:
+    case OperandOwnership::UnownedInstantaneousUse:
+    case OperandOwnership::BitwiseEscape:
+    // This Reborrow is actually the end of a nested borrow scope so handle it
+    // like a normal use of this guaranteed value.
+    case OperandOwnership::EndBorrow:
+      pushUse(use);
```
Comment (gottesmm) [inline lib/SIL/Utils/OwnershipUtils.cpp:904]: Why am I pushing things into usePoints that are already in usePoints? Shouldn't end_borrow be in the unreachable section and we continue with the other 3 cases?

File: lib/SIL/Utils/OwnershipUtils.cpp [hunk 5]
```diff
@@ -878,6 +845,160 @@ OwnedValueIntroducer swift::getSingleOwnedValueIntroducer(SILValue inputValue) {
...
+
+  // --- Transitively follow forwarded uses and look for escapes.
+
+  // FIXME: Remove this SmallPtrSet when destructures are reborrows. Currently
+  // it forwards multiple results! This means that usePoints could grow
+  // exponentially without a membership check. It's fine to do this membership
+  // check locally in this function (within a borrow scope). It isn't needed for
+  // the immediate uses, only the transitive uses.
+  SmallPtrSet<Operand *, 16> visitedUses;
+  auto pushUse = [&](Operand *use) {
```
Comment (gottesmm) [inline lib/SIL/Utils/OwnershipUtils.cpp:876]: Couldn't you make pushUse return true and just pass it into some of the cases below (see BorrowingOperand API use below).

File: lib/SIL/Utils/OwnershipUtils.cpp [hunk 6]
```diff
@@ -878,6 +845,160 @@ OwnedValueIntroducer swift::getSingleOwnedValueIntroducer(SILValue inputValue) {
...
+            for (auto *transitiveUse : transitiveValue->getUses()) {
+              pushUse(transitiveUse);
+            }
+            return true;
+          });
+      break;
+
+    case OperandOwnership::Borrow: {
+      BorrowingOperand(use).visitExtendedScopeEndingUses(
+        [&](Operand *endScope) {
```
Comment (gottesmm) [inline lib/SIL/Utils/OwnershipUtils.cpp:931]: pushUse with a small tweak could just be passed in directly here.

**Review by gottesmm — COMMENTED** [review]
Some comments.

**Review by gottesmm — APPROVED** [review]
I actually think some of my comments are with respect to #35671. Please fix them there. Beyond the small nits for the clang-format/addition of !=, this LGTM.

**Comment by atrick** [comment]
@swift-ci smoke test

**Comment by atrick** [comment]
@swift-ci test Windows

**Pull request merged**